designcompose/src/main/java/com/android/designcompose/FrameRender.kt

```diff
@@ -451,18 +459,30 @@ internal fun ContentDrawScope.render(
         )
 
     val customFillBrushFunction = customizations.getBrushFunction(name)
-    val customFillBrush =
+    var customFillBrush =
         if (customFillBrushFunction != null) {
             customFillBrushFunction()
         } else {
             customizations.getBrush(name)
         }
+    if (customFillBrush == null) {
+        view.shader.getOrNull()?.let {
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
```

Comment:
So this means shaders will only work on T or higher, and if not on T it'll just not render? Is this something new or is this a restriction that we've had before? Is this documented anywhere?

I don't like the idea of the doc rendering inaccurately if you're on too old of a build. IMO, the appropriate way to handle this would be to increase our minSDK version (after discussion with Ralph of course) or to seperate the shader parts into a seperate library with a higher minSDK, so that trying to build for a lower SDK causes a build failure, rather than an inaccurate rendering that looks like a bug.

Reply:
Yeah, the Android framework support for shaders landed in TM. I don't think there's a practical way to support them on SCv2 or earlier.

Reply:
https://developer.android.com/partners/google-apps/android-13/changes?hl=en#shaders

When shader isn't supported on old sdks, our render process will continue to render the background image if there is any. Currently the shader plugin will insert an image generated from the running shader when it writes the shader code to the plugin data.

Reply:
> Currently the shader plugin will insert an image generated from the running shader when it writes the shader code to the plugin data.

I don't think that's a good solution, I think any user would think that this were a bug. So @iamralpht it looks like we'll need to increase our minSDK to support this.

Reply:
I don't think we need to restrict SDK versions just for this feature. I would imagine there can be many users of DesignCompose that don't need shader support, and I don't think we need to restrict them to newer SDK versions. We can add documentation for this in our docs as well as the shader plugin UI.

Reply:
I vote not to restrict SDK too.

Reply:
@iamralpht any thoughts on restricting the SDK version?

Reply:
Ok, I have an alternate suggestion then, have the shader plugin default to placing an image that states something like "Shaders require Android 13(?) or greater", rather than just the static image. Probably include a checkbox, off by default, that allows the plugin user to instead use the static image of the shader.

Reply:
Ok. I added a line to the top of the shader plugin.
<img width="605" alt="Screenshot 2024-12-06 at 10 41 01" src="https://github.com/user-attachments/assets/4012f4f2-5d58-405f-a827-f66ca4b3103c">

Reply:
I specifically meant in my comment for it to default to putting a static image with that text into the file. My concern is that a designer will see that and not know what it means, or it won't be known that some systems are on the older SDK.

Reply:
I got that. The generated image can be used for lower sdk versions(although still need to improve the resolution). But if we put that text, it will not be good to use.

I personally think add this text to the plugin UI is informative.

Reply:
How will a tester who has no access to Figma and cannot see the plugin UI know that part of the design will not render accurately on certain test systems?

Reply:
The image fill of the figma node can be replaced, there is no guarantee that the generated image will be used. Even we add the text to the image, tester may still not be able to see it.

I think users of the figma plugin and DesignCompose especially users of this feature should be aware of the limitations. And of course we should document it properly in our documents.

Reply:
I agree with folks that increasing min sdk is too much for this case since shaders might be used by only a small fraction of our users. One way of solving this is to break design compose into modules where each module has its own min sdk. However, even that wouldn't solve the main problem for the developers who want shader for devices that support it while they can live with not having it on older devices (which is probably the majority of our users who want shaders).

That said, there is a risk in assuming what the designers want in this case. By falling back to a screenshot of the shader we lose the users who want to use shaders but don't want a static image of the shader on older android versions (imagine a shader that has a weird shape on the first frame). On the other hand, simply dropping the shader can also be confusing and makes it hard for users to use shaders in any meaningful way when they also need to deploy their apps on S-. 

My 2cents here is to give users the option of using static image while making it obvious what happens when shader is not supported. 
For example, is it possible to add an image field as the default image if the shader is not supported?
If I remember correctly, shaders can also be disabled at the hardware level, so just checking for the Android version may not be enough.

In terms of testing, I agree that it complicates the testing process since now systems with shaders have to be tested twice once with support and once without. However, it is the reality of Android. Despite all the compat libraries, app developers should still test their apps on different versions of Android.

wdyt?

Reply:
As we discussed last meeting, I added a fallback color using a color picker when shader isn't supported on lower sdk versions.

<img width="604" alt="Screenshot 2024-12-11 at 20 03 09" src="https://github.com/user-attachments/assets/22037d4b-4df4-4bef-b2bb-54cbf9d309b9" />

In the shader plugin UI, it has a checkbox to use that fallback color or the figma node image. Let me know if we want the checkbox or not(i.e. always using the fallback color as the fill).

-----
Regarding if shader can be disabled at hardware level. I didn't find any document stating that. Since AGSL is using skia engine which supports both cpu render and gpu render, the hardware acceleration may affect how shader renders but there seems no way to turn it off.

-----
Please take a look. Thanks!

Reply:
Couple comments on the UI:
- It isn't clear to me that the "Set Shader" button will save the shader into the node. Can we maybe have another little info hover popup that explains what it's going to do?
- Typo on "Andorid"
- I would separate the sentence "Only works on Android 13 and above" with the checkbox. I think the checkbox should be associated with the "Use a solid color..." sentence. Or you could just have the checkbox with a sentence like "Use a solid color when shaders are not supported", and then. have the Android 13 info inside the info hover.

Reply:
In addition to Rob's comment:
* I think "Set Shader" should be changed to "Apply". It should be in the lower right hand corner, where confirmation buttons are normally found.
* I would phrase the Android 13 bit to something like "Note: Shaders require Android 13+"


Honestly, I think it'd be best to not offer a choice about the fallback. I just realized that if we do offer the ability to fall back to a static image of the shader then we'd have the situation where designers may want a specific frame of the shader to be used, instead of the first frame. So we'd have feature requests to add controls for that. Offering a choice of color is also problematic, it makes it seem like it's a design choice, and it'll defeat the purpose of the fallback if someone chooses a color that doesn't clearly indicate that the design isn't rendering accurately. 

I'm still in favor of using the [standard purple/black checkerboard](https://www.reddit.com/r/explainlikeimfive/comments/7o3nuq/eli5_why_do_games_use_a_purple_checkerboard_for/) but I think just rendering a solid, bright color is also fine.

Reply:
BTW, I'm sorry you're having to deal with our design feedback. @iamralpht this is a situation where having a designer to help would have been really helpful.

Reply:
So what happens if the checkbox isn't checked? The text doesn't say what it does.

Reply:
There is info button on right side of the checkbox explaining that if unchecked it will use the image from the figma node.